README.md
# swagger-proxy [![Build Status](https://travis-ci.org/gchaincl/swagger-proxy.svg?branch=master)](https://travis-ci.org/gchaincl/swagger-proxy)
Swagger Proxy ensure HTTP Responses correctness based on swagger specs. 

# Usage
SwaggerProxy is designed to assist the development process, it can be used as a reverse proxy or as a middleware.

## Reverse Proxy
SwaggerProxy sits between the client/integration suite and the server as follows:
```
[Client] ---> [SwaggerProxy] ---> [Server]
```
SwaggerProxy will rely the request and analize the server response.

### Install
```bash
go get github.com/gchaincl/swagger-proxy/cmd/swagger-proxy`
```
### Run
```bash
$ swagger-proxy -h
Usage of swagger-proxy:
  -bind string
        Bind Address (default ":1234")
  -spec string
        Swagger Spec (default "swagger.yml")
  -target string
        Target (default "http://localhost:4321")
  -verbose
        Verbose
```

## Middleware
If your server is built in Golang, you can use it as a middleware:
```go
package main

import (
	"log"
	"net/http"

	proxy "github.com/gchaincl/swagger-proxy"
	"github.com/go-openapi/loads"
)

func main() {
	doc, err := loads.Spec("swagger.json")
	if err != nil {
		log.Fatal(err)
	}

	p, err := proxy.New(doc.Spec(), &proxy.LogReporter{}, proxy.WithVerbose(true))
	if err != nil {
		log.Fatal(err)
	}

	app := func(w http.ResponseWriter, req *http.Request) {
		if req.Method == "POST" {
			w.WriteHeader(201)
		}
	}
	log.Printf("Server Running")
	http.ListenAndServe(":8989", p.Handler(http.HandlerFunc(app)))
}

```


CHANGELOG.md
#Changelog

## v0.1.0 (Unreleased)

* Display pending operations when proxy shuts down
* Live spec reloading [\#1](https://github.com/gchaincl/swagger-proxy/issues/1)

## v0.0.1 (2017-05-25)

* Handle Shutdown properly
* Partial Header Validation
* Integration test
* Verifies `Produces` against Content-Type

## v0.0.0 (???)
* Initial proof of concept

proxy_test.go
package proxy

import (
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	"github.com/go-openapi/loads"
	"github.com/go-openapi/spec"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

type testResponse struct {
	status int
	header http.Header
	body   []byte
}

func (t *testResponse) Status() int         { return t.status }
func (t *testResponse) Header() http.Header { return t.header }
func (t *testResponse) Body() []byte        { return t.body }

type testReporter struct {
	success  []*http.Request
	errors   []error
	warnings []string
}

func (t *testReporter) Success(req *http.Request) {
	t.success = append(t.success, req)
}

func (t *testReporter) Error(req *http.Request, err error) {
	t.errors = append(t.errors, err)
}

func (t *testReporter) Warning(req *http.Request, msg string) {
	t.warnings = append(t.warnings, msg)
}

func (t *testReporter) Report() {
	panic("not implemented")
}

func openFixture(t *testing.T, name string) *spec.Swagger {
	doc, err := loads.Spec("./fixtures/" + name)
	require.NoError(t, err)
	return doc.Spec()
}

func TestValidation(t *testing.T) {
	spec := openFixture(t, "petstore.json")
	app, err := New(spec, nil)
	require.NoError(t, err)

	op := spec.Paths.Paths["/pet/findByStatus"].Get
	resp := &testResponse{
		header: http.Header{},
		body:   []byte("[]"),
	}
	resp.Header().Set("Content-Type", "application/json")
	for _, status := range []int{200, 400} {
		resp.status = status
		assert.NoError(t, app.Validate(resp, op))
	}

	resp.status = 999
	assert.Error(t, app.Validate(resp, op))
}

func TestHeaderValidation(t *testing.T) {
	swagger := openFixture(t, "petstore.json")
	proxy, err := New(swagger, nil)
	require.NoError(t, err)

	op := swagger.Paths.Paths["/user/login"].Get
	resp := &testResponse{
		status: 200,
		header: http.Header{},
	}

	t.Run("Valid Headers", func(t *testing.T) {
		resp.Header().Set("X-Rate-Limit", "32")
		resp.Header().Set("X-Expires-After", time.Now().Format(time.RFC3339))

		assert.NoError(t, proxy.ValidateHeaders(resp, op))
	})

	t.Run("Invalid Headers", func(t *testing.T) {
		resp.Header().Set("X-Rate-Limit", "NaN")
		resp.Header().Set("X-Expires-After", "Not a date-time")

		assert.Error(t, proxy.ValidateHeaders(resp, op))
	})
}

func TestProducesDefinition(t *testing.T) {
	swagger := openFixture(t, "petstore.json")
	app, err := New(swagger, nil)
	require.NoError(t, err)

	resp := &testResponse{header: http.Header{}}
	op := swagger.Paths.Paths["/pet"].Post
	assert.Error(t, app.ValidateMIME(resp, op),
		"Content-Type: application/json is missing")

	resp.Header().Set("Content-Type", "application/json")
	assert.NoError(t, app.ValidateMIME(resp, op),
		"Content-Type: application/json is present")

	t.Run("IsInheritedFromRoot", func(t *testing.T) {
		// MOVE `Produces` property to root
		swagger.Produces = []string{"application/json"}
		op.Produces = []string{}

		resp.Header().Del("Content-Type")
		assert.Error(t, app.ValidateMIME(resp, op),
			"Content-Type: application/json is missing")

		resp.Header().Set("Content-Type", "application/json")
		assert.NoError(t, app.ValidateMIME(resp, op),
			"Content-Type: application/json is present")
	})
}

func TestHTTPHandler(t *testing.T) {
	swagger := openFixture(t, "petstore.json")
	reporter := &testReporter{}
	app, err := New(swagger, reporter)
	require.NoError(t, err)

	fn := func(w http.ResponseWriter, req *http.Request) {
		if req.URL.String() == "/v2/store/inventory" {
			w.Header().Set("Content-Type", "application/json")
			w.Write([]byte("{}"))
		}
	}

	srv := httptest.NewServer(app.Handler(http.HandlerFunc(fn)))
	defer srv.Close()

	for _, url := range []string{
		"/v2/store/inventory",   // SUCCESS
		"/v2/pet/findByStatus",  // ERROR
		"/not_a_registered_url", // WARNING
	} {
		http.Get(srv.URL + url)
	}

	assert.Equal(t, 1, len(reporter.success))
	assert.Equal(t, 1, len(reporter.errors))
	assert.Equal(t, 1, len(reporter.warnings))
}

func TestPendingOperations(t *testing.T) {
	swagger := openFixture(t, "petstore.json")
	app, err := New(swagger, &testReporter{})
	require.NoError(t, err)

	// Dummy server using Handler as middleware
	srv := httptest.NewServer(app.Handler(
		http.HandlerFunc(
			func(w http.ResponseWriter, req *http.Request) {
			},
		),
	))
	defer srv.Close()

	pending := len(app.PendingOperations())

	// Count defines responses
	var opsCounter int
	WalkOps(swagger, func(path, meth string, op *spec.Operation) {
		opsCounter += 1
	})
	require.Equal(t, opsCounter, pending)

	t.Run("AreRemoved", func(t *testing.T) {
		http.Get(srv.URL + "/v2/pet/findByStatus")
		assert.Equal(t, pending-1, len(app.PendingOperations()))

		http.Post(srv.URL+"/v2/pet", "", nil)
		assert.Equal(t, pending-2, len(app.PendingOperations()))

		pending = len(app.PendingOperations())
	})

	t.Run("AreNotRemovedWhenPathNotFound", func(t *testing.T) {
		http.Get(srv.URL + "/not_and_endpoint")
		assert.Equal(t, pending, len(app.PendingOperations()))
	})

	t.Run("AreRemovedOnceWhenCalledNTimes", func(t *testing.T) {
		http.Get(srv.URL + "/v2/store/inventory")
		http.Get(srv.URL + "/v2/store/inventory")
		http.Get(srv.URL + "/v2/store/inventory")
		assert.Equal(t, pending-1, len(app.PendingOperations()))
	})
}


proxy.go
package proxy

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/http/httputil"
	"net/url"

	"github.com/go-openapi/errors"
	"github.com/go-openapi/spec"
	"github.com/go-openapi/strfmt"
	"github.com/go-openapi/swag"
	"github.com/go-openapi/validate"
	"github.com/gorilla/mux"
)

type Proxy struct {
	// Opts
	target  string
	verbose bool

	router       *mux.Router
	routes       map[*mux.Route]*spec.Operation
	reverseProxy http.Handler

	reporter Reporter

	doc               interface{} // This is useful for validate (TODO: find a better way)
	spec              *spec.Swagger
	pendingOperations map[*spec.Operation]struct{}
}

type ProxyOpt func(*Proxy)

func WithTarget(target string) ProxyOpt { return func(proxy *Proxy) { proxy.target = target } }
func WithVerbose(v bool) ProxyOpt       { return func(proxy *Proxy) { proxy.verbose = v } }

func New(s *spec.Swagger, reporter Reporter, opts ...ProxyOpt) (*Proxy, error) {
	proxy := &Proxy{
		target:   "http://localhost:8080",
		router:   mux.NewRouter(),
		routes:   make(map[*mux.Route]*spec.Operation),
		reporter: reporter,
	}

	if err := proxy.SetSpec(s); err != nil {
		return nil, err
	}

	for _, opt := range opts {
		opt(proxy)
	}

	rpURL, err := url.Parse(proxy.target)
	if err != nil {
		return nil, err
	}
	proxy.reverseProxy = httputil.NewSingleHostReverseProxy(rpURL)

	proxy.router.NotFoundHandler = http.HandlerFunc(proxy.notFound)
	proxy.registerPaths()

	return proxy, nil
}

func (proxy *Proxy) SetSpec(spec *spec.Swagger) error {
	// validate.NewSchemaValidator requires the spec as an interface{}
	// That's why we Unmarshal(Marshal()) the document
	data, err := json.Marshal(spec)
	if err != nil {
		return err
	}

	var doc interface{}
	if err := json.Unmarshal(data, &doc); err != nil {
		return err
	}

	proxy.doc = doc
	proxy.spec = spec

	proxy.registerPaths()
	return nil
}

func (proxy *Proxy) Router() http.Handler {
	return proxy.router
}

func (proxy *Proxy) Target() string {
	return proxy.target
}

func (proxy *Proxy) registerPaths() {
	proxy.pendingOperations = make(map[*spec.Operation]struct{})
	base := proxy.spec.BasePath

	router := mux.NewRouter()
	WalkOps(proxy.spec, func(path, method string, op *spec.Operation) {
		newPath := base + path
		if proxy.verbose {
			fmt.Printf("Register %s %s\n", method, newPath)
		}
		route := router.Handle(newPath, proxy.newHandler()).Methods(method)
		proxy.routes[route] = op
		proxy.pendingOperations[op] = struct{}{}
	})
	*proxy.router = *router
}

func (proxy *Proxy) notFound(w http.ResponseWriter, req *http.Request) {
	proxy.reporter.Warning(req, "Route not defined on the Spec")
	proxy.reverseProxy.ServeHTTP(w, req)
}

func (proxy *Proxy) newHandler() http.Handler {
	return proxy.Handler(proxy.reverseProxy)
}
func (proxy *Proxy) Handler(next http.Handler) http.Handler {
	fn := func(w http.ResponseWriter, req *http.Request) {
		wr := &WriterRecorder{ResponseWriter: w}
		next.ServeHTTP(wr, req)

		var match mux.RouteMatch
		proxy.router.Match(req, &match)
		op := proxy.routes[match.Route]

		if match.Handler == nil || op == nil {
			proxy.reporter.Warning(req, "Route not defined on the Spec")
			// Route hasn't been registered on the muxer
			return
		}
		proxy.operationExecuted(op)

		if err := proxy.Validate(wr, op); err != nil {
			proxy.reporter.Error(req, err)
		} else {
			proxy.reporter.Success(req)
		}
	}
	return http.HandlerFunc(fn)
}

type validatorFunc func(Response, *spec.Operation) error

func (proxy *Proxy) Validate(resp Response, op *spec.Operation) error {
	if _, ok := op.Responses.StatusCodeResponses[resp.Status()]; !ok {
		return fmt.Errorf("Server Status %d not defined by the spec", resp.Status())
	}

	var validators = []validatorFunc{
		proxy.ValidateMIME,
		proxy.ValidateHeaders,
		proxy.ValidateBody,
	}

	var errs []error
	for _, v := range validators {
		if err := v(resp, op); err != nil {
			if cErr, ok := err.(*errors.CompositeError); ok {
				errs = append(errs, cErr.Errors...)
			} else {
				errs = append(errs, err)
			}
		}
	}

	if len(errs) == 0 {
		return nil
	}
	return errors.CompositeValidationError(errs...)
}

func (proxy *Proxy) ValidateMIME(resp Response, op *spec.Operation) error {
	// Use Operation Spec or fallback to root
	produces := op.Produces
	if len(produces) == 0 {
		produces = proxy.spec.Produces
	}

	ct := resp.Header().Get("Content-Type")
	if len(produces) == 0 {
		return nil
	}

	for _, mime := range produces {
		if ct == mime {
			return nil
		}
	}

	return fmt.Errorf("Content-Type Error: Should produce %q, but got: '%s'", produces, ct)
}

func (proxy *Proxy) ValidateHeaders(resp Response, op *spec.Operation) error {
	var errs []error

	r := op.Responses.StatusCodeResponses[resp.Status()]
	for key, spec := range r.Headers {
		if err := validateHeaderValue(key, resp.Header().Get(key), &spec); err != nil {
			errs = append(errs, err)
		}
	}

	if len(errs) == 0 {
		return nil
	}

	return errors.CompositeValidationError(errs...)
}

func (proxy *Proxy) ValidateBody(resp Response, op *spec.Operation) error {
	r := op.Responses.StatusCodeResponses[resp.Status()]
	if r.Schema == nil {
		return nil
	}

	var data interface{}
	if err := json.Unmarshal(resp.Body(), &data); err != nil {
		return err
	}

	v := validate.NewSchemaValidator(r.Schema, proxy.doc, "", strfmt.Default)
	if result := v.Validate(data); result.HasErrors() {
		return result.AsError()
	}

	return nil
}

func validateHeaderValue(key, value string, spec *spec.Header) error {
	if value == "" {
		return fmt.Errorf("%s in headers is missing", key)
	}

	// TODO: Implement the rest of the format validators
	switch spec.Format {
	case "int32":
		_, err := swag.ConvertInt32(value)
		return err
	case "date-time":
		_, err := strfmt.ParseDateTime(value)
		return err
	}
	return nil
}

func (proxy *Proxy) PendingOperations() []*spec.Operation {
	var ops []*spec.Operation
	for op, _ := range proxy.pendingOperations {
		ops = append(ops, op)
	}
	return ops
}

func (proxy *Proxy) operationExecuted(op *spec.Operation) {
	delete(proxy.pendingOperations, op)
}

type WalkOpsFunc func(path, meth string, op *spec.Operation)

func WalkOps(spec *spec.Swagger, fn WalkOpsFunc) {
	for path, props := range spec.Paths.Paths {
		for meth, op := range getOperations(&props) {
			fn(path, meth, op)
		}
	}
}

func getOperations(props *spec.PathItem) map[string]*spec.Operation {
	ops := map[string]*spec.Operation{
		"DELETE":  props.Delete,
		"GET":     props.Get,
		"HEAD":    props.Head,
		"OPTIONS": props.Options,
		"PATCH":   props.Patch,
		"POST":    props.Post,
		"PUT":     props.Put,
	}

	// Keep those != nil
	for key, op := range ops {
		if op == nil {
			delete(ops, key)
		}
	}

	return ops
}


reporter.go
package proxy

import (
	"fmt"
	"net/http"

	"github.com/fatih/color"
	"github.com/go-openapi/errors"
)

type Reporter interface {
	Success(req *http.Request)
	Error(req *http.Request, err error)
	Warning(req *http.Request, msg string)
	Report()
}

type LogReporter struct {
}

func (r *LogReporter) Success(req *http.Request) {
	fmt.Fprintf(color.Output, "%s %s %s\n",
		color.GreenString("✔"), req.Method, req.URL,
	)
}

func (r *LogReporter) Error(req *http.Request, err error) {
	fmt.Fprintf(color.Output, "%s %s %s\n",
		color.RedString("✗"), req.Method, req.URL,
	)
	if cErr, ok := err.(*errors.CompositeError); ok {
		for i, err := range cErr.Errors {
			fmt.Printf("  %d) %s\n", i+1, err)
		}

	} else {
		fmt.Printf("  => %s\n", err)
	}
}

func (r *LogReporter) Warning(req *http.Request, msg string) {
	fmt.Fprintf(color.Output, "%s %s %s\n",
		color.YellowString("!"), req.Method, req.URL,
	)
	fmt.Printf("  WARNING: %s\n", msg)
}

func (r *LogReporter) Report() {}


cmd/swagger-proxy/main.go
package main

import (
	"flag"
	"fmt"
	"log"
	"net/http"
	"os"
	"os/signal"
	"path/filepath"
	"time"

	"github.com/fsnotify/fsnotify"
	proxy "github.com/gchaincl/swagger-proxy"
	"github.com/go-openapi/loads"
)

const version = "v0.0.1"

func serve(proxy *proxy.Proxy, bind string) error {
	s := http.Server{
		Addr:    bind,
		Handler: proxy.Router(),
	}

	errC := make(chan error)
	go func() {
		log.Println("SwaggerProxy", version, "listening on", bind, "->", proxy.Target())
		errC <- s.ListenAndServe()
	}()

	sigC := make(chan os.Signal, 1)
	signal.Notify(sigC, os.Interrupt)

	select {
	case err := <-errC:
		return err
	case s := <-sigC:
		log.Printf("%s", s)
		return nil
	}
}

func sameFiles(a, b string) bool {
	absA, err := filepath.Abs(a)
	if err != nil {
		return false
	}

	absB, err := filepath.Abs(b)
	if err != nil {
		return false
	}

	return absA == absB
}

func reload(proxy *proxy.Proxy, spec string) error {
	doc, err := loads.Spec(spec)
	if err != nil {
		return err
	}

	if err := proxy.SetSpec(doc.Spec()); err != nil {
		return err
	}
	return nil
}

func watchFor(px *proxy.Proxy, spec string) error {
	watcher, err := fsnotify.NewWatcher()
	if err != nil {
		return err
	}

	if err := watcher.Add(filepath.Dir(spec)); err != nil {
		return err
	}
	defer watcher.Close()

	needReload := false
	for {
		select {
		case <-time.After(100 * time.Millisecond):
			if needReload {
				needReload = false
				log.Println("Reloading", spec)
				if err := reload(px, spec); err != nil {
					log.Println(err)
					continue
				}
				needReload = false
			}
		case ev := <-watcher.Events:
			if !sameFiles(ev.Name, spec) {
				continue
			}

			if ev.Op != fsnotify.Write && ev.Op != fsnotify.Chmod {
				continue
			}

			needReload = true
		case <-watcher.Errors:
			// TODO: handle this
		}
	}
}

func main() {
	bind := flag.String("bind", ":1234", "Bind Address")
	spec := flag.String("spec", "swagger.yml", "Swagger Spec")
	target := flag.String("target", "http://localhost:4321", "Target")
	verbose := flag.Bool("verbose", false, "Verbose")
	flag.Parse()

	doc, err := loads.Spec(*spec)
	if err != nil {
		log.Fatal(err)
	}

	proxy, err := proxy.New(doc.Spec(), &proxy.LogReporter{},
		proxy.WithTarget(*target),
		proxy.WithVerbose(*verbose),
	)
	if err != nil {
		log.Fatal(err)
	}

	go watchFor(proxy, *spec)

	if err := serve(proxy, *bind); err != nil {
		log.Println(err)
	}

	// Report PendingOperations
	fmt.Println("Pending Operations:")
	fmt.Println("------------------")
	for i, op := range proxy.PendingOperations() {
		fmt.Printf("%03d) id=%s\n", i+1, op.ID)
	}
}
